### PR TITLE
Refactor k12

### PR DESCRIPTION
**Pull request by aewag**
The k12 crate had its own keccak-p permutation implementation. 
~With https://github.com/RustCrypto/sponges/pull/7 a generic keccak-p implementation would be available, which could be used within the k12 crate.~
With https://github.com/RustCrypto/hashes/pull/458 TurboSHAKE is available, which can be used within the `k12` crate.

`k12` uses a block size of 8192 to split the input. The maximum supported block size of `block-buffer` is limited to 256. Therefore this implementation buffers the input within an additional array.

### CODE REVIEW

File: k12/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ categories = ["cryptography", "no-std"]
 
 [dependencies]
 digest = { version = "0.10", features = ["alloc"] }
+keccak = { git = "https://github.com/aewag/sponges", branch = "1-add-keccak-p" }
```
Comment (aewag) [inline k12/Cargo.toml]: This needs to be changed in the future. `keccak-p` is not yet available in the official keccak crate.

**Comment by gvanas** [comment]
For K12, it would be interesting to also have SIMD implementations of parallel Keccak-p permutations.

Sorry, I cannot help directly because I am not fluent in Rust, but I can at least point out such implementations in [XKCP/K12](https://github.com/XKCP/K12) for inspiration.

**Comment by aewag** [comment]
> For K12, it would be interesting to also have SIMD implementations of parallel Keccak-p permutations.
> 
> Sorry, I cannot help directly because I am not fluent in Rust, but I can at least point out such implementations in [XKCP/K12](https://github.com/XKCP/K12) for inspiration.

Yes, that definitely would get some nice improvements. Currently I have made a draft PR for SIMD backed Keccak-f permutations in https://github.com/RustCrypto/sponges/pull/8. As soon as this is available, the k12 implementation could be updated to incorporate it.

EDIT: https://github.com/RustCrypto/sponges/pull/8 would need an generic SIMD backed keccak-p permutation for use within k12. I'll extend the draft PR.

**Comment by gvanas** [comment]
I was not aware of that PR. Nice!

**Comment by tarcieri** [comment]
@aewag this is great work. Anything I can do to help move it along?

**Comment by aewag** [comment]
> @aewag this is great work. Anything I can do to help move it along?

If you have time to review https://github.com/RustCrypto/sponges/pull/7, that would be great. Other than that, this propably needs a rebase and should more or less work.

(I planned to further update the implementation to be alloc-free and maybe also in the future with SIMD support, but I will not able to work on this in the near future.)

EDIT: I just rebased this on top of the current master.

**Comment by tarcieri** [comment]
@aewag went ahead and merged https://github.com/RustCrypto/sponges/pull/7. We can probably cut another release of the `keccak` crate.

I wanted to point out this is probably the biggest problem with the `k12` crate in its current form:

https://github.com/RustCrypto/hashes/blob/1cdbd5363f8e20cc378e19bce4d6470143f08ff7/k12/src/lib.rs#L37-L39

It buffers all of the input in a `Vec<u8>` and doesn't actually hash it until `XofReader::read`:

https://github.com/RustCrypto/hashes/blob/1cdbd5363f8e20cc378e19bce4d6470143f08ff7/k12/src/lib.rs#L128-L177

That makes the implementation unusable for large inputs, and it doesn't properly implement the `XofReader` contract in that it will panic if `XofReader::read` is called repeatedly.

**Comment by aewag** [comment]
> @aewag went ahead and merged [RustCrypto/sponges#7](https://github.com/RustCrypto/sponges/pull/7). We can probably cut another release of the `keccak` crate.
> 
> I wanted to point out this is probably the biggest problem with the `k12` crate in its current form.
> That makes the implementation unusable for large inputs, and it doesn't properly implement the `XofReader` contract in that it will panic if `XofReader::read` is called repeatedly.

Yep, agreed. I started to work on these (two) issue(s), but I don't have yet a working implementation.

**Comment by aewag** [comment]
@tarcieri I finished the refactoring. The PR is ready for review.

File: k12/src/lib.rs [hunk 2]
```diff
@@ -16,242 +11,215 @@
...
+
+use sha3::{TurboShake128, TurboShake128Core, TurboShake128ReaderCore};
+
+const CHUNK_SIZE: usize = 8192;
+const CHAINING_VALUE_SIZE: usize = 32;
+const LENGTH_ENCODE_SIZE: usize = 255;
+
+#[doc = "Core "]
+#[doc = "KangarooTwelve"]
+#[doc = " hasher state."]
```
Comment (tarcieri) [inline k12/src/lib.rs]:
```suggestion
/// Core [`KangarooTwelve`] hasher state.
```

Comment (aewag) [reply k12/src/lib.rs]: Thanks :+1: I fixed the comments.

**Review by tarcieri — APPROVED** [review]

**Comment by aewag** [comment]
Thank you! :+1:

**Comment by tarcieri** [comment]
I can cut a release if you'd like

**Comment by aewag** [comment]
Yeah, that would be nice. Thanks!

**Pull request merged**